dask/dataframe/core.py

```diff
@@ -351,6 +353,57 @@ def __init__(self, dsk, name, meta, divisions):
         self._meta = meta
         self.divisions = tuple(divisions)
 
+        # Optionally cast object dtypes to `pyarrow` strings
+        if dask.config.get("dataframe.convert_string"):
+            try:
+                import pyarrow  # noqa: F401
+            except ImportError:
+                raise RuntimeError(
+                    "Using dask's `dataframe.convert_string` configuration "
+                    "option requires `pyarrow` to be installed."
+                )
+            if not PANDAS_GT_130:
+                raise RuntimeError(
+                    "Using dask's `dataframe.convert_string` configuration "
+                    "option requires pandas>=1.3.0 to be installed. "
+                    f"pandas={str(PANDAS_VERSION)} is currently using used."
+                )
+
+            from dask.dataframe._pyarrow import (
+                is_object_string_dataframe,
+                is_object_string_dtype,
+                is_object_string_index,
+                is_object_string_series,
+                to_pyarrow_string,
+            )
+
+            if (
+                is_object_string_dataframe(meta)
+                or is_object_string_series(meta)
+                or is_object_string_index(meta)
+            ):
+                # Prior to pandas=1.4, `pd.Index` couldn't contain extension dtypes.
+                # Here we don't cast objects to pyarrow strings where only the index
+                # contains non-pyarrow string data.
+                if not PANDAS_GT_140 and (
```

Comment:
I think I'd rather move this to `def to_pyarrow_string(df)`. If what we're converting is an Index, and `not PANDAS_GT_140`, return as-is.

Reply:
I'm not sure that that will work as we need to gate the `result = self.map_partitions(to_pyarrow_string)` call below from happening when only `Index` objects contain non-pyarrow text data. Otherwise, we end up in an infinite recursion

Reply:
I see what you mean, good point, nevermind.